### PR TITLE
LGW: Lookup node IPs on every NP/LB/EIP service event

### PR DESCRIPTION
**Pull request by astoycos**
This Commit ensures that the correct IPtables rules are added for 
externalIP services in the case that a secondary node IP was added 
after ovnkube-node had already been started.

It simply looks up the node ips for each service event rather than
storing the ips in memeory.

This PR is in response to the [following bug](https://bugzilla.redhat.com/show_bug.cgi?id=1959798)

### CODE REVIEW

**Comment by astoycos** [comment]
Probably need to figure out how to Mock the Net package here... Unless I move the test to e2e tests

**Comment by alexanderConstantinescu** [comment]
I have not really reviewed this, but @trozet added an `addressManager` on the node side during his work on commit: https://github.com/ovn-org/ovn-kubernetes/pull/2159/commits/5b913613ffdd94d6366326f7e062f550508b2714 This tracks all node IPs. Could we make use of the fact that we constantly react and track state w.r.t any newly added IP address, here?

**Comment by astoycos** [comment]
/assign @JacobTanenbaum 

I basically followed how you Mocked DNS calls however I did not have add additional wrapping layer in the utils package like you did in `util/dns.go`

In the future I probably will but for now I think this is the simplest way  forward

**Comment by astoycos** [comment]
@alexanderConstantinescu Yeah that was the path I first went down.. However the added complexity was a bunch, especially since this is for LGW 

For example... Essentially we would need an "add Node IP " event to Trigger an "addService" event which would either 
1. Require us to track all services in an in memory structure
or  
2. Write a new "Iptables" controller that could respond to events from both the AddService and AddNodeIP events 

Just recalculating node IPs on any service event seemed like the simplest way forward for the attached bug

**Comment by alexanderConstantinescu** [comment]
> @alexanderConstantinescu Yeah that was the path I first went down.. However the added complexity was somewhat large, especially since this is for LGW
> 
> For example... Essentially we would need an "add Node IP " event to Trigger an "addService" event which would either
> 
> 1. Require us to track all services in an in memory structure
>    or
> 2. Write a new "Iptables" controller that could respond to events from both the AddService and AddNodeIP events
> 
> Just recalculating node IPs on any service event seemed like the simplest way forward for the attached bug

I might be missing some details, but I don't fully follow. 

Essentially, what you're describing suggest that we should handle the cases:

- "1) user first modifies the service and 2) modifies the IP address on the node to reflect the service changes he did in 1)" 
- "1) user first modifies the IP address on the node and 2) modifies the service to reflect the changes he did in 1)". 

and thus we would need to update the external IP setup whenever: a) user modifies the service and b) when a user adds a new IP to the node which is associated with the external IP. However, your PR is only doing a). It's not reacting to node IP address modifications currently, and subsequently updating any previously made service setup. 

And given that you're only doing a) I think it's safe to say that we could just always run `addressManager` (instead of only in shared gateway mode as we do today: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_init.go#L225) and access `addressManager.addresses` whenever we react to a service add/delete instead of listing all the IP addresses on the node by doing `net.InterfaceAddrs`. The only reason I mention this is because it would be faster. `addressManager.addresses` is a `sets.String` (i.e: map) which will give us a O(1) access, where as doing `net.InterfaceAddrs` is more costly. Also you wouldn't need to mock `net.InterfaceAddrs` then. 

However, if this PR is needed because things are broken on older versions and we don't want to backport `addressManager`, then fine, that's a sufficiently valid argument I guess.

**Comment by alexanderConstantinescu** [comment]
> However, if this PR is needed because things are broken on older versions and we don't want to backport `addressManager`, then fine, that's a sufficiently valid argument I guess.

Actually, forget my comments. I realize now that it is quite complex to do as it would require a backport of specific bits of the `addressManager` (for ex: _not_ backporting any master side code which reacts to the node annotations set by `addressManager`). We would only need the updates `addressManager` does to its internal data, without any of the additional stuff it does.

**Comment by aojea** [comment]
@astoycos let me suggest you a different approach, we should run away of mocks if possible, is not a best practices for golang testing, however dependency injection is.
So, we just need to get the addresses on the node and "fake" them on the tests, you can just inject a dependency (we don't earn nothing asserting "I called a function" , what we want is to generate tables of testing "getlocal return 1 IPv4, 10 ipv6, a mix of them, ...)

```
type localPortWatcher struct {
	recorder    record.EventRecorder
	gatewayIPv4 string
	gatewayIPv6 string
        getLocalAddr func() (map[string]net, error)
}
```

this way in your test you can use ,
```
fakeLocalPortWatcher  = &localPortWatcher{
   recorder fakeRecored
   getLocalAddr myFakeLocalAddrFunc()
```

Also, I'd like you to check why we need a `map[string]net` as data structure, if we only care about ips maybe we just need a string or a net.IP

**Comment by astoycos** [comment]
Thanks for the review @aojea! Yeah that makes sense I can do that, after learning how to implement mocking I do think it adds a considerable bit of complexity/overhead to the code. 

Granted if we move away from mocking as a whole I think it needs to be brought up in the community meeting since in some places we rely on it pretty heavily.

**Comment by aojea** [comment]
> Thanks for the review @aojea! Yeah that makes sense I can do that, after learning how to implement mocking I do think it adds a considerable bit of complexity/overhead to the code.
> 
> Granted if we move away from mocking as a whole I think it needs to be brought up in the community meeting since in some places we rely on it pretty heavily.

If you mean about nbctl mocking, there is no other way imho, so that is ok, but if we move to bindings, we have much more room to improve ...

**Comment by astoycos** [comment]
Yeah NBCTL is tough but I was talking more about how we mocking with DNS package calls like so 
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/mocks/DNSOps.go
Which is what I followed originally for this PR.

**Comment by aojea** [comment]
> Yeah NBCTL is tough but I was talking more about how we mocking with DNS package calls like so
> https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/mocks/DNSOps.go
> Which is what I followed originally for this PR.

sorry,  I didn't see that, just to be clear, mocking is not the devil and something that is wrong, but different languages has different ways of doing things, and idiomatic ways are always preferred, better long term supports, sustainable, ... 
what we should avoid is global variables though ... :grimacing:

**Comment by astoycos** [comment]
Agreed, Also for 

> Also, I'd like you to check why we need a map[string]net as data structure, if we only care about ips maybe we just need a string or a net.IP

We use a map because for application of Iptables rules We have to 
1. Check if the EIP is a nodeIP 
```
if _, exists := l.localAddrSet[externalIP]; exists {
					iptRules = append(iptRules, getExternalIPTRules(port, externalIP, ip)...)
					klog.V(5).Infof("Will add iptables rule for ExternalIP: %s", externalIP)
```
2.  See if the EIP is routable 
```
} else if l.networkHasAddress(net.ParseIP(externalIP)) {
					klog.V(5).Infof("ExternalIP: %s is reachable through one of the interfaces on this node, will skip setup", externalIP)
```
3. If not 1 or 2 there is no gateway for the EIP 

If you just keep track of the net you lose the original node IP, and if you keep track of just the IP you use lose the mask information. 

So Instead we could just use a set to check if the external IP is equal to a node address and  an array of nets to keep track of the reachable networks.  However I don't _think_ this would be any better.

File: go-controller/pkg/node/gateway_localnet_linux_test.go [hunk 1]
```diff
@@ -24,14 +26,16 @@ const (
 	v6localnetGatewayIP = "fd00:96:1::1"
 )
 
-func getFakeLocalAddrs() map[string]net.IPNet {
-	localAddrSet := make(map[string]net.IPNet)
+func getFakeIpNet() []net.Addr {
```
Comment (JacobTanenbaum) [inline go-controller/pkg/node/gateway_localnet_linux_test.go:29]: nit: golang naming conventions Initialism should be the same case `getFakeIPNet()` 

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

**Comment by astoycos** [comment]
@JacobTanenbaum Antonio wanted me to move away from Mocking for this and instead do some dependency injection see #2229 

Can you let me know what you think there?

**Comment by JacobTanenbaum** [comment]
I agree that there needs to be a decision at the next community meeting as to if we want to continue using mocks I was unaware there was any debate

**Comment by astoycos** [comment]
I can leave both these PRs up as examples Since it's pretty easy to understand what's going on here

**Comment by aojea** [comment]
> I agree that there needs to be a decision at the next community meeting as to if we want to continue using mocks I was unaware there was any debate.
> 
> I don't know what the best solution is

I didn't mean to open a debate about this, just that in this case is clear that mocks are not the best solution, it is double of LOCs and adds a dependency, compared to injecting a function https://github.com/ovn-org/ovn-kubernetes/pull/2226

**Pull request closed without merging**